components/script/dom/readablestream.rs

```diff
@@ -108,426 +255,705 @@ impl ReadableStream {
-            ExternalUnderlyingSource::FetchResponse => (vec![], false),
-        };
-        ExternalUnderlyingSourceController {
-            buffer: RefCell::new(buffer),
-            closed: Cell::new(false),
-            in_memory: Cell::new(in_mem),
+    /// <https://streams.spec.whatwg.org/#readable-stream-get-num-read-requests>
+    pub fn get_num_read_requests(&self) -> usize {
+        assert!(self.has_default_reader());
+        match self.reader {
+            ReaderType::Default(ref reader) => {
+                let reader = reader
+                    .get()
+                    .expect("Stream must have a reader when get num read requests is called into.");
+                reader.get_num_read_requests()
+            },
+            ReaderType::BYOB(_) => unreachable!(
+                "Stream must have a default reader when get num read requests is called into."
+            ),
         }
     }
 
-    /// Does the stream have all data in memory?
-    pub fn in_memory(&self) -> bool {
-        self.in_memory.get()
+    /// <https://streams.spec.whatwg.org/#readable-stream-fulfill-read-request>
+    pub fn fulfill_read_request(&self, chunk: SafeHandleValue, done: bool) {
+        // step 1 - Assert: ! ReadableStreamHasDefaultReader(stream) is true.
+        assert!(self.has_default_reader());
+        match self.reader {
+            ReaderType::Default(ref reader) => {
+                // step 2 - Let reader be stream.[[reader]].
+                let reader = reader
+                    .get()
+                    .expect("Stream must have a reader when a read request is fulfilled.");
+                // step 3 - Assert: reader.[[readRequests]] is not empty.
+                assert_ne!(reader.get_num_read_requests(), 0);
+                // step 4 & 5
+                // Let readRequest be reader.[[readRequests]][0]. & Remove readRequest from reader.[[readRequests]].
+                let request = reader.remove_read_request();
+
+                if done {
+                    // step 6 - If done is true, perform readRequest’s close steps.
+                    request.close_steps();
+                } else {
+                    // step 7 - Otherwise, perform readRequest’s chunk steps, given chunk.
+                    let result = RootedTraceableBox::new(Heap::default());
+                    result.set(*chunk);
+                    request.chunk_steps(result);
+                }
+            },
+            ReaderType::BYOB(_) => unreachable!(
+                "Stream must have a default reader when fulfill read requests is called into."
+            ),
+        }
     }
 
-    /// Return bytes synchronously if the stream has all data in memory.
-    pub fn get_in_memory_bytes(&self) -> Option<Vec<u8>> {
-        if self.in_memory.get() {
-            return Some(self.buffer.borrow().clone());
+    /// <https://streams.spec.whatwg.org/#readable-stream-close>
+    pub fn close(&self) {
+        // step 1
```

Comment:
Since we're using the actual sentences elsewhere, we can be consistent and do it here too.